lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+)
+
+const suffixLength = 8
+
+// Sorted is a disk-spilling container, that sorts added values by the
+// accompanying keys.
+// You must not add new values after calling PopFirst/-Last.
+type Sorted interface {
+	Add(k []byte, v Value)
+	PopFirst(v Value) bool
+	PopLast(v Value) bool
+	NewIterator(reverse bool) Iterator
+	Bytes() int
+	Items() int
+	SetOverflowBytes(bytes int)
+	Close()
+}
+
+type sorted struct {
+	commonSorted
+	base
+}
+
+type keyIterator interface {
+	Iterator
+	key() []byte
+}
+
+type commonSorted interface {
+	common
+	add(k []byte, v Value)
+	getFirst(v Value) bool
+	getLast(v Value) bool
+	dropFirst(bytes int)
+	dropLast(bytes int)
+	newIterator(p iteratorParent, reverse bool) keyIterator
+}
+
+// NewSorted returns an implementaiton of Sorted, spilling to disk at location.
+func NewSorted(location string) Sorted {
+	o := &sorted{base: newBase(location)}
+	o.commonSorted = &memorySorted{}
+	return o
+}
+
+func (o *sorted) Add(k []byte, v Value) {
+	if o.startSpilling(o.Bytes() + v.Bytes()) {
```

Comment:
This (and the other similar ones) should probably be a separate `o.maybeSpill()`